### PR TITLE
Detailed settings for performance critical code higlightings

### PR DESCRIPTION
**Pull request by krasnotsvetov**

### CODE REVIEW

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/PerformanceCriticalCodeStageUtil.cs [hunk 1]
```diff
@@ -135,6 +135,25 @@ public static bool IsNullComparisonWithUnityObject(IEqualityExpression equalityE
             return false;
         }
 
+        public static bool HasPerformanceSensitiveAttribute(IFunctionDeclaration functionDeclaration)
+        {
+            return HasSpecificAttribute(functionDeclaration, "PerformanceCharacteristicsHintAttribute");
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/PerformanceCriticalCodeStageUtil.cs]: I think this name only works if we're giving it an argument that specifies what the characteristics are. I'll add a top level comment about attribute names

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/Highlightings/PerformanceHighlightingAttributeIds.cs [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
 {
     // Note that all attribute IDs should start with "ReSharper Unity " to appear properly in both Rider and ReSharper
     public static class PerformanceHighlightingAttributeIds
     {
         public const string CAMERA_MAIN = "ReSharper Unity Expensive Camera Main Usage";
         public const string COSTLY_METHOD_INVOCATION = "ReSharper Unity Expensive Method Invocation";
         public const string NULL_COMPARISON = "ReSharper Unity Expensive Null Comparison";
-        public const string COSTLY_METHOD_HIGHLIGHTER = "ReSharper Unity Performance Critical Context";
+        public const string PERFORMANCE_CRITICAL_CONTEXT_METHOD_HIGHLIGHTER = "ReSharper Unity Performance Critical Context";
+        public const string COSTLY_METHOD_CONTEXT_HIGHLIGHTER = "ReSharper Unity Performance Critical Line Marker";
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/Highlightings/PerformanceHighlightingAttributeIds.cs]: This gets displayed to the user, but not sure what it means - if it's the gutter line marker for the current method, do we need a new highlight attribute, or just hide/show it based on context?

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/Highlightings/UnityHighlightingAbstractStage.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            IEnumerable<IUnityDeclarationHiglightingProvider> higlightingProviders, UnityApi api,
+            UnityHighlightingContributor unityHighlightingContributor, DaemonProcessKind processKind) : base(process, file)
+        {
+            myHiglightingProviders = higlightingProviders;
+            myAPI = api;
+            myUnityHighlightingContributor = unityHighlightingContributor;
+            myProcessKind = processKind;
+
+            myEventFunctions = DaemonProcess.CustomData.GetData(UnityEventFunctionAnalyzer.UnityEventFunctionNodeKey)
+                ?.ToJetHashSet();
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/CSharp/Daemon/Stages/Highlightings/UnityHighlightingAbstractStage.cs]: How do we know this has been populated by the analyser first?

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/Analysis/UnityEventFunctionAnalyzer.cs [hunk 4]
```diff
@@ -106,14 +103,27 @@ public UnityEventFunctionAnalyzer(UnityApi unityApi, UnityImplicitUsageHighlight
...
-                            myImplicitUsageHighlightingContributor.AddUnityImplicitHighlightingForEventFunction(consumer, method, function);
+                            AddIcon(method, data);
                             AddMethodSignatureInspections(consumer, method, function, candidate.Match);
                         }
                     }
                 }
             }
         }
 
+        private void AddIcon(IMethod method, ElementProblemAnalyzerData data)
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/CSharp/Daemon/Stages/Analysis/UnityEventFunctionAnalyzer.cs]: Is this name correct? It's not adding an icon

File: resharper/resharper-unity/src/Rider/ConnectionTracker.cs [hunk 5]
```diff
@@ -22,6 +22,9 @@ public ConnectionTracker(Lifetime lifetime, ILogger logger, UnityHost host, Unit
         {
             unitySolutionTracker.IsUnityProjectFolder.AdviseOnce(lifetime, args =>
             {
+                if (!args)
```
[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/Rider/ConnectionTracker.cs]: We will have crash in tests, also before this change we periodically check Unity Editor for not unity solutions

File: resharper/resharper-unity/src/Rider/Highlightings/PerformanceCriticalCodeContextHighlighter.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            var settingsStore = psiDocumentRangeView.GetSettingsStore();
+            
+            if (!settingsStore.GetValue((UnitySettings key) => key.EnablePerformanceCriticalAnalysis))
+                return;
+
+            if (settingsStore.GetValue((UnitySettings key) => key.EnableLineMarkerForPerformanceCriticalCode))
+                return;
+            
+            if (!settingsStore.GetValue((UnitySettings key) => key.EnableLineMarkerForActivePerformanceCriticalMethod))
```
[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/Rider/Highlightings/PerformanceCriticalCodeContextHighlighter.cs]: Thanks, I forgot to replace it with combo box, will do it :)

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/Highlightings/UnityHighlightingAbstractStage.cs (hunk 3 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/Highlightings/UnityHighlightingAbstractStage.cs]: This stage executed after CSharpErrorStage

File: resharper/resharper-unity/src/Settings/UnitySettings.cs [hunk 7]
```diff
@@ -32,5 +29,19 @@ public class UnitySettings
 
         [SettingsEntry(true, "Enables syntax error highlighting, brace matching and more of YAML files for Unity")]
         public bool IsYamlParsingEnabled;
+
+        // Analysis
+        [SettingsEntry(true, "Enables performance analysis in frequently called code")]
+        public bool EnablePerformanceCriticalAnalysis;
```
[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/Settings/UnitySettings.cs]: No, I will return back. Thanks!

**Pull request merged**